contracts/pools/PoolCollection.sol

```diff
@@ -1497,48 +1612,64 @@ contract PoolCollection is IPoolCollection, Owned, BlockNumber, Utils {
             d: liquidity.baseTokenTradingLiquidity
         });
 
-        Fraction112 memory averageRate = averageRateInfo.rate;
+        Fraction112 memory rate = averageRates.rate;
+
+        if (!spotRate.isPositive() || !rate.isPositive()) {
+            return PoolRateState.Uninitialized;
+        }
+
+        Fraction memory invSpotRate = spotRate.inverse();
+        Fraction112 memory invRate = averageRates.invRate;
 
-        if (!spotRate.isPositive() || !averageRate.isPositive()) {
+        if (!invSpotRate.isPositive() || !invRate.isPositive()) {
             return PoolRateState.Uninitialized;
         }
 
-        if (averageRateInfo.blockNumber != _blockNumber()) {
-            averageRate = _calcAverageRate(averageRate, spotRate);
+        if (averageRates.blockNumber != _blockNumber()) {
+            rate = _calcAverageRate(rate, spotRate);
+            invRate = _calcAverageRate(invRate, invSpotRate);
         }
 
-        if (MathEx.isInRange(averageRate.fromFraction112(), spotRate, RATE_MAX_DEVIATION_PPM)) {
+        if (
+            MathEx.isInRange(rate.fromFraction112(), spotRate, RATE_MAX_DEVIATION_PPM) &&
+            MathEx.isInRange(invRate.fromFraction112(), invSpotRate, RATE_MAX_DEVIATION_PPM)
+        ) {
             return PoolRateState.Stable;
         }
 
         return PoolRateState.Unstable;
     }
 
     /**
-     * @dev updates the average rate
+     * @dev updates the average rates
      */
-    function _updateAverageRate(Pool storage data, Fraction memory spotRate) private {
+    function _updateAverageRates(Pool storage data, Fraction memory spotRate) private {
         uint32 blockNumber = _blockNumber();
 
-        if (data.averageRate.blockNumber != blockNumber) {
-            data.averageRate = AverageRate({
+        if (data.averageRates.blockNumber != blockNumber) {
+            data.averageRates = AverageRates({
                 blockNumber: blockNumber,
-                rate: _calcAverageRate(data.averageRate.rate, spotRate)
+                rate: _calcAverageRate(data.averageRates.rate, spotRate),
+                invRate: _calcAverageRate(data.averageRates.invRate, spotRate.inverse())
             });
         }
     }
 
     /**
      * @dev calculates the average rate
      */
-    function _calcAverageRate(Fraction112 memory averageRate, Fraction memory spotRate)
+    function _calcAverageRate(Fraction112 memory averageRate, Fraction memory poolRate)
```

Comment:
Semantic: I don't think that we should use `poolRate` here when we use `spotRate` and `SPOT_RATE` everywhere else.

Reply:
`spotRate` is a bad name because it can be both the spot and the inverse spot rate. I'll change it to `rate`, but in general - please refrain from such nitpicking on tightly scheduled PRs.